frontend/catalyst/programs/verification.py

```diff
@@ -0,0 +1,209 @@
+    for op in tape.operations:
+        state = op_checker_fn(op, state)
+        if has_nested_tapes(op):
+            for region in nested_quantum_regions(op):
+                if region.trace is not None:
+                    with EvaluationContext.frame_tracing_context(ctx, region.trace):
+                        state = _verify_nested(region.quantum_tape, state, op_checker_fn)
+                else:
+                    state = _verify_nested(region.quantum_tape, state, op_checker_fn)
+    return state
+
+
+EMPTY_PROPERTIES = OperationProperties(False, False, False)
+
+
+@transform
+def verify_no_state_variance_returns(tape: QuantumTape) -> None:
+    """Verify that no measuremnts contain state or variance."""
+
+    if any(isinstance(m, (StateMP, VnEntropyMP, MutualInfoMP)) for m in tape.measurements):
+        raise DifferentiableCompileError("State returns are forbidden in gradients")
+
+    if any(isinstance(m, VarianceMP) for m in tape.measurements):
+        raise DifferentiableCompileError("Variance returns are forbidden in gradients")
+
+    return (tape,), lambda x: x[0]
+
+
+@transform
+def verify_operations(tape: QuantumTape, grad_method, qjit_device):
+    """verify the quantum program against Catalyst requirements. This transform makes no
+    transformations.
+
+    Raises:
+        DifferentiableCompileError: gradient-related error
+        CompileError: compilation error
+    """
+
+    # FIXME: How should we re-organize the code to avoid this kind of circular dependency?
+    from catalyst.api_extensions import MidCircuitMeasure
+    from catalyst.api_extensions.quantum_operators import Adjoint, QCtrl
+    from catalyst.jax_tracer import HybridOp
+
+    def _paramshift_op_checker(op):
+        if not isinstance(op, HybridOp):
+            if op.grad_method not in {"A", None}:
+                raise DifferentiableCompileError(
+                    f"{op.name} does not support analytic differentiation"
+                )
+
+    def _mcm_op_checker(op):
+        if isinstance(op, MidCircuitMeasure):
+            raise DifferentiableCompileError(f"{op.name} is not allowed in gradinets")
+
+    def _adj_op_checker(op):
+        if not qjit_device.qjit_capabilities.native_ops.get(
+            op.name, EMPTY_PROPERTIES
+        ).differentiable:
+            raise DifferentiableCompileError(
+                f"{op.name} is non-differentiable on '{qjit_device.original_device.name}' device"
+            )
+
+    def _ctrl_op_checker(op, in_qctrl, in_controllable):
+        if not (in_qctrl or in_controllable):
+            # PennyLane has many flavors of controlled operations. Here we check if the operation is
+            # supported as a self-contained native operation for a device.
+            if op.__class__ in {Controlled, ControlledOp, ControlledQubitUnitary}:
+                if not qjit_device.qjit_capabilities.native_ops.get(op.name):
+                    return _ctrl_op_checker(op.base, in_qctrl, in_controllable=True)
```

Comment:
I'm trying to understand why it wouldn't be sufficient to perform the same thing as with the inverse below. Or why the inverse doesn't need the same method as here 🤔 

Let's take `Adjoint(CRZ)` for instance. The adjoint check
```
op_name = op.base.name if isinstance(op, Controlled) else op.name
if not qjit_device.qjit_capabilities.native_ops.get(op_name, ...).invertible:
```
will verify wether `RZ` is invertible, rather than `CRZ`.

The control check will enter this branch, detect we have a ControlledOp, and then verify if `CRZ` is natively supported. If it's not natively supported it will verify whether `RZ` is supported and controllable.

- So if `CRZ` is supported Adjoint verified the wrong thing.
- The alternative check whether RZ is controllable seems to assume that we are actually mapping `CRZ` to `RZ` somewhere, which as far as I know is not the case. (Although I did throw together a PR earlier that does exactly that, but at the decomp layer, so pre verification: https://github.com/PennyLaneAI/catalyst/pull/792). If these assumptions hold then I think we need to reject `CRZ` at this stage even if `RZ.controllable` is true.

Reply:
Let me comment this. We can encode quantum control of a gate using two "dimensions" : 1) We can put the gate into `QCtrl` region of Catalyst 2) We can wrap the gate with `Controllable` class of PL. Also we should mind "named-controllable" gates like `CNOT`, that is why we check both `Controllable` and `ControlledOp`. Obviously, we should expect to see both things at once.

Notably, both ways of the encoding are recursive, so we can have `QCtrl(QCtrl( ... ))` and `Controllable(Controllable(...))`.

The above verification code uses two boolean variables `in_qctrl` and `in_controllable` to guide the recursion. It is something that we do not need when we deal with Adjoint, because afaik we only have one way of encoding it.

Reply:
Updated the check in the `check_inv` function to look at `op.__class__ in {Controlled, ControlledOp, ControlledQubitUnitary}` instead of `isinstance(op, Controlled)`